enhancements/management-workload-partitioning.md

```diff
@@ -0,0 +1,787 @@
+   file for kubelet to enable the same workload partition. The file
+   should only be readable by the kubelet.
+4. The user runs the installer to create the standard manifests, adds
+   their extra manifests from steps 2 and 3, then creates the cluster.
+5. The kubelet starts up and finds the configuration file enabling the
+   new feature.
+6. The kubelet advertises `cpu.workload.openshift.io/management`
+   extended resources on the node based on the number of CPUs in the
+   host.
+7. The kubelet reads static pod definitions. It replaces the `cpu`
+   requests with `cpu.workload.openshift.io/management` requests of
+   the same value and adds the
+   `io.openshift.workload.management.cpu/{container-name}` annotations
+   for CRI-O with the same values.
+8. Something schedules a regular pod with the
+   `workload.openshift.io/management` annotation in a namespace with
+   the `workload.openshift.io/allowed=management` label.
+9. The admission hook modifies the pod, replacing the CPU requests
+   with `cpu.workload.openshift.io/management` requests and adding the
+   `io.openshift.workload.management.cpu/{container-name}` annotations
+   for CRI-O.
+10. The scheduler sees the new pod and finds available
+    `cpu.workload.openshift.io/management` resources on the node. The
+    scheduler places the pod on the node.
+11. Repeat steps 8-10 until all pods are running.
+12. Single-node deployment comes up with management components
+    constrained to subset of available CPUs.
+
+#### Workload Annotation
+
+The `workload.openshift.io` annotation on each pod needs to allow us
+to add new parameters in the future, so the value will be a
+struct. Initially, it will encode 2 values.
+
+The `target` for the workloads is in the annotation name to make it
+easy for CRI-O and other lower-level components to detect it without
+understanding the full JSON struct in the value.
+
+The `effect` field will control whether the request is a soft or hard
+rule. It can contain either `PreferredDuringScheduling` for soft
+requests or `RequiredDuringScheduling` for hard requests. Initially we
+will only implement the soft request mode and that will be the
+default.
+
+```yaml
+metadata:
+  annotations:
+    workload.openshift.io/management: |
+      {"effect": "PreferredDuringScheduling"}
+```
+
+The admission hook will reject pods with multiple annotations, for
+now. Future work may allow multiple targets with different priorities
+to support clusters with different types of configurations.
+
+#### Pod mutation
+
+The kubelet and API admission hook will change pods annotated with
+`workload.openshift.io/management` so the CPU requests are replaced
+with management CPU requests and an annotation is added with the same
+value.
+
+```yaml
+requests:
+  cpu:
+    400m
+```
+
+becomes
+
+```yaml
+requests:
+  cpu.workload.openshift.io/management: 400
+```
+
+and
+
+```yaml
+annotations:
+  io.openshift.workload.management.cpu/{container-name}: 400
+```
```

Comment:
We can have a table of conversions here. E.g. 400m -> 400, 0.3 -> 300 1 -> 1000 and so on.